README.md
fretcalc v 0.17

GNU/GPL copying rules apply, see COPYING

INTRODUCTION:
 fretcalc is a simple console fret calculator for Linux.
 
COMPILATION:
 Go into fretcalc source directory and type 'make'.
 It should compile without errors.

INSTALLATION:
 Currently running make will produce a single binary in the source
 directory. You may put it anywhere you want. 

RUNNING:
 fretcalc -n <num_frets> -s <scale_length>

OPTIONS: 
 Options:
    -a <fret1,fret2,...fretx) additional half frets for dulcimer - use with -d
       half frets allowed: 0,1,3,4,6 and multiples of 7 for each 
       (7,8,10,11,13,etc.) 
    -c <frets_per_octave>     custom number of frets per octave
    -d, --dulcimer            calculate for dulcimer
    -h, --help                print this help
    -n <num_frets>            number of frets
    -o, --output              write output to calc.out
    -p <num_decimal_places>   precision of calculation display (1-5, default 3)
    -s <scale_length>         scale length
    -V, --verbose             be verbose
    -v, --version             print version number
    -w, --warranty            show warranty and copyright

FILES
 ./.fretcalc        Users fretcalc configuration.
    n=<num_frets> - required
    s=<scale_length> - required	
    p=<precision> - optional, default 3
    c=<frets_per_octave> - optional, default 12
    d=1 - optional (dulcimer)
    a=<fret1,fret2,...fretx> - optional (half frets)


ChangeLog
v 0.17 (Jan 15, 2005)
  + Added FretCalculator::setDulcHalfFrets - pass vector
    containing half fret numbers to this method.
  + Added 'd' (dulcimer) and 'a' (additional half frets) to
    default input file (.fretcalc)

v 0.16 (Jan 15, 2005)
  + Added 'p' (precision) and 'c' (custom frets per octave) 
    default input file (.fretcalc)

v 0.15 (Jan 12, 2005)
  + Separated declaration and implementation of FretCalculator
    class into FretCalculator.h and FretCalculator.cpp
  + Modified Makefile
  + Moved flag validation to mutator methods

v 0.14 (Jan 10, 2005)
  + Added default configuration file ability (.fretcalc)

v 0.13 (Jan 06, 2005)
  + Changed int variables to bool
    + print_help 
    + print_version
    + print_warranty
    + verbose
    + is_dulcimer
    + out_file
    + s_found
    + n_found
  + Changed location of loop counter (i) declaration in param_init()

v 0.12 (Jan 05, 2005)
  + Code now OO - FretCalcultor class

v 0.11 (Dec 28, 2004)
  + added iomanip

v 0.10 (Dec 28, 2004)
  + Changed code from C to C++

v 0.09 (Sep 29, 2004)
  + Added -c (custom) - number of frets per octave which allows
    non 12-tone calcuations. Non dulcimer calc only.

v 0.08 (Sep 25, 2004)
  + fixed bug dulc_calc that caused half frets 7 and above to not calculate 
  + -o will now write half frets to output file

v 0.07 (Sep 24, 2004)
  + -a now accpets multiple half frets separated by commas
  + -a now accepts 0+ (-a 0)

v 0.06 (Sep 24, 2004)
  + Added -a option which allows user to request half fret for dulcimer

v 0.05 (Sep 22, 2004)
  + Fixed bug in -p (was settting num_frets_val)
  + Check that argument value isn't another argument flag

v 0.04 (Sep 22, 2004)
  + Added dulc_calc() function and -d argument
  + Renamed calc_frets() to frets_calc()
  + Removed verbose print statements from write_calc()
 
v 0.03 (Sep 22, 2004)
  + Added MAX_FRETS (100)
  + Added calc_frets() function

v 0.02 (Sep 21, 2004)
  + Added ChangeLog
  + Added -p arg for precision of decimal.

v 0.01 (Sep 20, 2004) 
  + Initial Release



fretcalc.cpp
// fretcalc.cpp
#include "FretCalculator.h"
#include <iostream>
#include <iomanip>
#include <string>
#include <fstream>
#include <cmath>
#include <vector>

#include <stdio.h>
#include <stdlib.h>
#include <cstring>
using namespace std;

// vars
vector <int> dulc_half_frets(0);

// function delcarations
void param_init(int, char *[]);

// variable declarations
char * endptr;
double scale_length;
int num_frets;
int frets_per_octave;
int precision;
bool print_help     = false;
bool print_version  = false;
bool print_warranty = false;
bool verbose        = false;
bool is_dulcimer    = false;
bool out_file       = false;
bool s_found        = false;
bool n_found        = false;

int main (int argc, char **argv)
{

  // check for parameters
  param_init(argc,argv);

  // create object
  FretCalculator FC;

  // check for message flags
  if(print_help)
  {
    FC.printHelp();
    exit(0);
  } 
  else if(print_warranty)
  {
    FC.printVersion();
    FC.printWarranty();
    exit(0);
  }
  else if(print_version)
  {
    FC.printVersion();
    exit(0);
  }

  // Calculate if num_frets and scale_length
  // Remove this if statement to use
  // num_frets and scale_length defaults
  if(!n_found && !s_found)
  {
    // check for .fretcalc file 
    ifstream inFile(".fretcalc");
    if(inFile)
    {
      while(! inFile.eof())
      {
        string input_str;
        string flag_name;
        string flag_value;
        unsigned int pos;

        getline(inFile,input_str);
        pos = input_str.find("=");
        if(pos != string::npos)
        {
          flag_name.assign(input_str,pos-1,1);  // assumes 1 character flagname
          flag_value.assign(input_str,pos+1,input_str.size()-pos-1); 
          if( flag_name.compare("n") == 0 )
          {
            //cout << "n flag found (" << flag_value << ")" << endl;
            n_found = true;
            num_frets = atoi(flag_value.c_str());
          }
          if( flag_name.compare("s") == 0 )
          {
            //cout << "s flag found (" << flag_value << ")" << endl;
            s_found = true;
            scale_length = atof(flag_value.c_str());
          }
          if( flag_name.compare("p") == 0 )
          {
            //cout << "p flag found (" << flag_value << ")" << endl;
            precision = atoi(flag_value.c_str());
          }
          if( flag_name.compare("c") == 0 )
          {
            //cout << "c flag found (" << flag_value << ")" << endl;
            frets_per_octave = atoi(flag_value.c_str());
          }
          if( flag_name.compare("d") == 0 )
          {
            //cout << "d flag found (" << flag_value << ")" << endl;
            is_dulcimer = true;
          }
          if( flag_name.compare("a") == 0 )
          {
            // split on commas
            char *pch;       // pointer for tokens
            const char *ptr1 = 0;
            int len = flag_value.length();
            char *ptr2 = new char[len + 1]; // including null
            // copy characters out of sting into allocated memory
            flag_value.copy(ptr2,len,0);
            ptr2[len] = 0; // add null terminator
            pch = strtok(ptr2,",");
            while (pch != NULL)
            {
              int half_fret = atoi(pch);
              if( half_fret >= 0 )
              {
                dulc_half_frets.push_back(half_fret);
                pch = strtok(NULL,",");
              }
            }
          }
        }
      }
      inFile.close();
    } 
  }

  if(n_found && s_found )
  {

    // set all params
    FC.setScaleLength(scale_length);
    FC.setNumFrets(num_frets);
    FC.setFretsPerOctave(frets_per_octave);
    FC.setPrecision(precision);
    FC.setDulcHalfFrets(dulc_half_frets);

    // do fret calculation
    if( is_dulcimer )
      FC.calcDulcFrets();
    else
      FC.calcFrets();

    if(out_file)
    {
     FC.writeFrets();
      if(verbose)
        FC.printFrets();
    } 
    else
    {
      FC.printFrets();
    }
  
  } else {
    FC.printHelp();
  }

  return 0;
}

void param_init(int argc,char *argv[])
{
  // first, get the parameters
  char *str;             // string pointer
  const char *sub = "-"; // substring pointer
  char *ptr;             // pointer for substring address
  char *pch;             // pointer for tokens
  int pos;               // element position for found substring

  for (int i = 1; i < argc; i++)
  {
    // additional half frets for dulcimer
    if (strcmp(argv[i], "-a") == 0)
    {
      if (i + 1 < argc)
      {
        // make sure "i+1" isn't another argument flag
        str = argv[i+1];
        ptr = strstr(str,sub);
        pos = (ptr - str);
        if( pos == 0 ) {
          continue;
        }

        // split on commas
        pch = strtok(argv[++i],",");
        while (pch != NULL)
        {
          int half_fret = atoi(pch);
          if( half_fret >= 0 )
          {
            dulc_half_frets.push_back(half_fret);
            pch = strtok(NULL,",");
          }
        }
      } else {
        // default
      }
    }

    // custom (custom) non-12 tone 
    if (strcmp(argv[i], "-c") == 0)
    {
      if (i + 1 < argc)
      {
        // make sure "i+1" isn't another argument flag
        str = argv[i+1];
        ptr = strstr(str,sub);
        pos = (ptr - str);
        if( pos != 0 )
        {
          frets_per_octave = atoi(argv[++i]);
        }
      }
    }

    // dulcimer
    if ((strcmp(argv[i], "-d") == 0) ||
        (strcmp(argv[i], "--d") == 0) ||
        (strcmp(argv[i], "-dulcimer") == 0) ||
        (strcmp(argv[i], "--dulcimer") == 0)) {
      is_dulcimer = true;
    }

    // help
    if ((strcmp(argv[i], "-h") == 0) ||
        (strcmp(argv[i], "--h") == 0) ||
        (strcmp(argv[i], "-help") == 0) ||
        (strcmp(argv[i], "--help") == 0)) {
      print_help = true;
    }

    // write to output file 'calc.out' 
    if ((strcmp(argv[i], "-o") == 0) ||
        (strcmp(argv[i], "--o") == 0) ||
        (strcmp(argv[i], "-output") == 0) ||
        (strcmp(argv[i], "--output") == 0)) {
      out_file = true;
    }

    // precision (precision)
    if (strcmp(argv[i], "-p") == 0)
    {
      if (i + 1 < argc)
      {
        // make sure "i+1" isn't another argument flag
        str = argv[i+1];
        ptr = strstr(str,sub);
        pos = (ptr - str);
        if( pos != 0 ) {
          precision = atoi(argv[++i]);
        }
      }
    }

    // verbose
    if ((strcmp(argv[i], "-V") == 0) ||
        (strcmp(argv[i], "--V") == 0) ||
        (strcmp(argv[i], "-verbose") == 0) ||
        (strcmp(argv[i], "--verbose") == 0)) {
      verbose = true;
    }

    // version and credits
    if ((strcmp(argv[i], "-v") == 0) ||
        (strcmp(argv[i], "--v") == 0) ||
        (strcmp(argv[i], "-version") == 0) ||
        (strcmp(argv[i], "--version") == 0)) {
      print_version = true;
    }

    // warranty
    if ((strcmp(argv[i], "-w") == 0) ||
        (strcmp(argv[i], "--w") == 0) ||
        (strcmp(argv[i], "-warranty") == 0) ||
        (strcmp(argv[i], "--warranty") == 0)) {
      print_version = true;
      print_warranty = true;
    }

    // number of frets (num_frets_val)
    if (strcmp(argv[i], "-n") == 0)
    {
      if (i + 1 < argc)
      {
        // make sure "i+1" isn't another argument flag
        str = argv[i+1];
        ptr = strstr(str,sub);
        pos = (ptr - str);
        if( pos != 0 )
        {
          n_found = true;
          num_frets = atoi(argv[++i]);
        }
      }
    }

    // scale length (scale_length) 
    if (strcmp(argv[i], "-s") == 0)
    {
      if (i + 1 < argc)
      {
        // make sure "i+1" isn't another argument flag
        str = argv[i+1];
        ptr = strstr(str,sub);
        pos = (ptr - str);
        if( pos != 0 )
        {
          s_found = true;
          scale_length = strtod(argv[++i],&endptr);
        }
      }
    }

  }
}


FretCalculator.h
// FretCalculator.h
#ifndef FRETCALCULATOR_H
#define FRETCALCULATOR_H
#include <vector>
const int MAX_FRETS = 100;

class FretCalculator
{
  public:
  // default constructor
  FretCalculator();
  FretCalculator(double, int);
  // destructor
  ~FretCalculator();
  // methods
  // frets
  void calcFrets();
  void calcDulcFrets();
  void printFrets();
  void writeFrets();
  // mutator
  void setScaleLength(double);
  void setNumFrets(int);
  void setFretsPerOctave(int);
  void setPrecision(int);
  void setDulcHalfFrets(const std::vector<int> &vec);
  // accessor
  double getScaleLength();
  int getNumFrets();
  int getFretsPerOctave();
  int getPrecision();
  // help/version/warranty
  void printHelp();
  void printVersion();
  void printWarranty();

  // data
  private:
    double scale_length;
    int num_frets;
    double nut_to_fret [MAX_FRETS];
    int frets_per_octave;
    int precision;
};

#endif // FRETCALCULATOR_H


FretCalculator.cpp
// FretCalculator.cpp
#include "FretCalculator.h"
#include <iostream>
#include <iomanip>
#include <string>
#include <fstream>
#include <cmath>
#include <vector>
using namespace std;

// constants
const string PROGRAM_NAME = "fretcalc";
const string VERSION_NUMBER = "0.17";

// vars
double dulc_frets [MAX_FRETS];
int dulc_half_frets_bool[MAX_FRETS]; 

FretCalculator::FretCalculator()
{
  // initialize object
  scale_length     = 25;
  num_frets        = 24;
  frets_per_octave = 12;
  precision        = 3;
}

FretCalculator::FretCalculator(double sl, int nf)
{
  // initialize object
  scale_length     = sl;
  num_frets        = nf;
  frets_per_octave = 12;
  precision        = 3;
}

FretCalculator::~FretCalculator()
{
}

void FretCalculator::calcFrets()
{
  int i;

  nut_to_fret[0] = 0.0;
  for (i = 1; i <= num_frets; i++)
  {
    double ii = i;
    double dist;
    dist = scale_length - (scale_length / pow(2,(ii/frets_per_octave)));
    nut_to_fret[i] = dist;
  }
}

void FretCalculator::calcDulcFrets()
{
  /* Half frets allowed: 0,1,3,4,6 and multiples of 7 (7,8,10,11,13,etc.)
     Half frets not allowed: 2,5 and multiples of 7 (9,12,etc.)

     Note Dulc_fret Guitar_fret
     --------------------------
      D     0  (7)    0  (12)
      D#    0+ (7+)   1  (13)
      E     1  (8)    2  (14)
      F     1+ (8+)   3  (15)
      F#    2  (9)    4  (16)
      G     3  (10)   5  (17)
      G#    3+ (10+)  6  (18)
      A     4  (11)   7  (19)
      A#    4+ (11+)  8  (20)
      B     5  (12)   9  (21)
      C     6  (13)   10 (22)
      C#    6+ (13+)  11 (23)
   */
  int i,j,k,dulc_half_fret_flag;
  int zero_counter;
  int one_counter;
  int three_counter;
  int four_counter;
  int six_counter;

  dulc_half_fret_flag = 0; /* calculate a "half" fret */
  zero_counter = 0;
  one_counter = 1;
  three_counter = 3;
  four_counter = 4;
  six_counter = 6;

  nut_to_fret[0] = 0.0;
  for (i = 1, j = 1; j <= num_frets; i++)
  {
    if( (i-1) % 12 == 0 )       /* Dulc(j): 0+,7+,14+ - Guitar(i): 1,13,25 */
    {
      if( dulc_half_frets_bool[zero_counter] )
      {
        k = zero_counter;
        zero_counter += 7;
        dulc_half_fret_flag = 1;
      }
      else
      {
        zero_counter += 7;
        continue;
      }
    }
    else if( (i-3) % 12 == 0 )  /* Dulc(j): 1+,8+,15+ - Guitar(i): 3,15,27 */
    {
      if( dulc_half_frets_bool[one_counter] )
      {
        k = one_counter;
        one_counter += 7;
        dulc_half_fret_flag = 1;
      }
      else
      {
        one_counter += 7;
        continue;
      }
    }
    else if( (i-4) % 12 == 0 )  /* Dulc(j): 2,9,16 - Guitar(i): 4,16,28 */
    {
      k = i - 2;
      if( k >= 0 ) // >= 2?
      {
        dulc_half_frets_bool[k] = 0;
        dulc_half_fret_flag = 0;
      }
    }
    else if( (i-6) % 12 == 0 )  /* Dulc(j): 3+,10+,17+ - Guitar(i): 6,18,30 */
    {
      if( dulc_half_frets_bool[three_counter] )
      {
        k = three_counter;
        three_counter += 7;
        dulc_half_fret_flag = 1;
      }
      else
      {
        three_counter += 7;
        continue;
      }
    } else if( (i-8) % 12 == 0 )  /* Dulc(j): 4+,11+,18+ - Guitar(i): 8,20,32 */    {
      if( dulc_half_frets_bool[four_counter] )
      {
        k = four_counter;
        four_counter += 7;
        dulc_half_fret_flag = 1;
      }
      else
      {
        four_counter += 7;
        continue;
      }
    }
    else if( (i-9) % 12 == 0 )  /* Dulc(j): 5,12,19 - Guitar(i): 9,21,33 */
    {
      k = i - 4;
      if( k >= 0 ) // >= 4?
      {
        dulc_half_frets_bool[k] = 0;
        dulc_half_fret_flag = 0;
      }
    }
    else if( (i-11) % 12 == 0 ) /* Dulc(j): 6+,13+,25+ - Guitar(i): 11,23,35 */
    {
      if( dulc_half_frets_bool[six_counter] )
      {
        k = six_counter;
        six_counter += 7;
        dulc_half_fret_flag = 1;
      }
      else
      {
        six_counter += 7;
        continue;
      }
    }

    double ii = i;
    double dist;
    dist = scale_length - (scale_length / pow(2,(ii/12)));
    if( dulc_half_fret_flag )
      dulc_frets[k] = dist;
    else
      nut_to_fret[j] = dist;

    if( dulc_half_fret_flag )
      dulc_half_fret_flag = 0;
    else
      j++;
  }
}

void FretCalculator::printFrets()
{
  int i;

  cout << "Number of Frets: "  << num_frets << endl;
  cout << "Scale Length: "     << scale_length << endl;
  cout << "Frets Per Octave: " << frets_per_octave << endl;

  // start from 0 since there may be a ducl_half_frets[0]
  for (i = 0; i <= num_frets; i++)
  {
    double dist;
    dist = nut_to_fret[i];


    // now setprecision() will set number of decimal places
    cout.setf(ios::fixed);

    /* print regular frets first */
    if( i > 0 )
    {
      if( precision == 1 )
        cout << "Fret " << i << ": " << setprecision(1) << dist << endl;
      else if (precision == 2 )
        cout << "Fret " << i << ": " << setprecision(2) << dist << endl;
      else if (precision == 3 )
        cout << "Fret " << i << ": " << setprecision(3) << dist << endl;
      else if (precision == 4 )
        cout << "Fret " << i << ": " << setprecision(4) << dist << endl;
      else if (precision == 5 )
        cout << "Fret " << i << ": " << setprecision(5) << dist << endl;
      else
        cout << "Fret " << i << ": " << setprecision(3) << dist << endl;
    }

    // print dulcimer frets second
    if( dulc_half_frets_bool[i] )
    {
      dist = dulc_frets[i];
    
      if( precision == 1 )
        cout << "Fret " << i << "+: " << setprecision(1) << dist << endl;
      else if (precision == 2 )
        cout << "Fret " << i << "+: " << setprecision(2) << dist << endl;
      else if (precision == 3 )
        cout << "Fret " << i << "+: " << setprecision(3) << dist << endl;
      else if (precision == 4 )
        cout << "Fret " << i << "+: " << setprecision(4) << dist << endl;
      else if (precision == 5 )
        cout << "Fret " << i << "+: " << setprecision(5) << dist << endl;
      else
        cout << "Fret " << i << "+: " << setprecision(3) << dist << endl;
      }
  }
}

void FretCalculator::writeFrets()
{
  int i;

  ofstream outfile("calc.out", ios::out);
  if(! outfile)
  {
    cout << "Error opening output file" << endl;
    return;
  }

  // file opened successfully
  cout << "File calc.out created" << endl;

  outfile << "Number of Frets: "  << num_frets << endl;
  outfile << "Scale Length: "     << scale_length << endl;
  outfile << "Frets Per Octave: " << frets_per_octave << endl;

  // now setprecision() will set number of decimal places
  outfile.setf(ios::fixed);


  // start from 0 since there may be a ducl_half_frets[0]
  for (i = 0; i <= num_frets; i++)
  {
    double dist;
    dist = nut_to_fret[i];

    // print regular frets first
    if( i > 0 )
    {
      if( precision == 1 )
        outfile << "Fret " << i << ": " << setprecision(1) << dist << endl;
      else if (precision == 2 )
        outfile << "Fret " << i << ": " << setprecision(2) << dist << endl;
      else if (precision == 3 )
        outfile << "Fret " << i << ": " << setprecision(3) << dist << endl;
      else if (precision == 4 )
        outfile << "Fret " << i << ": " << setprecision(4) << dist << endl;
      else if (precision == 5 )
        outfile << "Fret " << i << ": " << setprecision(5) << dist << endl;
      else
        outfile << "Fret " << i << ": " << setprecision(3) << dist << endl;
    }

    // print dulcimer frets second
    if( dulc_half_frets_bool[i] )
    {
      dist = dulc_frets[i];

      if( precision == 1 )
        outfile << "Fret: " << i << "+: " << setprecision(1) << dist << endl;
      else if (precision == 2 )
        outfile << "Fret: " << i << "+: " << setprecision(2) << dist << endl;
      else if (precision == 3 )
        outfile << "Fret: " << i << "+: " << setprecision(3) << dist << endl;
      else if (precision == 4 )
        outfile << "Fret: " << i << "+: " << setprecision(4) << dist << endl;
      else if (precision == 5 )
        outfile << "Fret: " << i << "+: " << setprecision(5) << dist << endl;
      else
        outfile << "Fret: " << i << "+: " << setprecision(3) << dist << endl;
      }
  }

  outfile.close();

}

void FretCalculator::setScaleLength(double sl)
{
  scale_length = sl;
}

void FretCalculator::setNumFrets(int nf)
{
  num_frets = nf;
}

void FretCalculator::setFretsPerOctave(int fpo)
{
  if( fpo <= 0 )
    fpo = 12;

  frets_per_octave = fpo;
}

void FretCalculator::setPrecision(int prec)
{
  if( prec <= 0 || prec > 5 )
    prec = 3;

  precision = prec;
}

void FretCalculator::setDulcHalfFrets(const vector<int> &vec)
{
  std::vector<int>::const_iterator p1;
  for( p1 = vec.begin(); p1 != vec.end(); p1++ )
  {
    //cout << *p1 << ' ';
    dulc_half_frets_bool[*p1] = 1;
  }
}

double FretCalculator::getScaleLength()
{
  return scale_length;
}

int FretCalculator::getNumFrets()
{
  return num_frets;
}

int FretCalculator::getFretsPerOctave()
{
  return frets_per_octave;
}

int FretCalculator::getPrecision()
{
  return precision;
}

void FretCalculator::printHelp()
{
  cout << "Usage: fretcalc -n <num_frets> -s <scale_length> [OPTIONS]" << endl;
  cout << "e.g: fretcalc -n 24 -s 25.5" << endl;
  cout << "e.g: fretcalc -d -a 6 -n 24 -s 25.5" << endl;
  cout << endl;
  cout << "Options:" << endl;
  cout << "  -a <fret1,fret2,...fretx) additional half frets for dulcimer - use with -d" << endl;
  cout << "     half frets allowed: 0,1,3,4,6 and multiples of 7 for each (7,8,10,11,13,etc.)" << endl;
  cout << "  -c <frets_per_octave>     custom number of frets per octave" << endl;
  cout << "  -d, --dulcimer            calculate for dulcimer" << endl;
  cout << "  -h, --help                print this help" << endl;
  cout << "  -n <num_frets>            number of frets" << endl;
  cout << "  -o, --output              write output to calc.out" << endl;
  cout << "  -p <num_decimal_places>   precision of calculation display (1-5, default 3)" << endl;
  cout << "  -s <scale_length>         scale length" << endl;
  cout << "  -V, --verbose             be verbose" << endl;
  cout << "  -v, --version             print version number" << endl;
  cout << "  -w, --warranty            show warranty and copyright" << endl;
}

void FretCalculator::printVersion()
{
  cout << PROGRAM_NAME << " " << VERSION_NUMBER << endl;
}

void FretCalculator::printWarranty()
{
  cout << "  Copyright (c) 1996--2005 by" << endl;
  cout << "  Douglas S Sparling <doug@dougsparling.com>" << endl;
  cout << endl;
  cout << "    This program is free software; you can redistribute it and/or" << endl;
  cout << "modify it under the terms of the GNU General Public License version 2" << endl;
  cout << "as published by the Free Software Foundation." << endl;
  cout << endl;
  cout << "    This program is distributed in the hope that it will be useful," << endl;
  cout << "but WITHOUT ANY WARRANTY; without even the implied warranty of" << endl;
  cout << "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU" << endl;
  cout << "General Public License for more details." << endl;
  cout << endl;
  cout << "    You should have received a copy (refer to the file COPYING) of the" <<  endl;
  cout << "GNU General Public License along with this program; if not, write to" << endl;
 cout << "the Free Software Foundation, Inc., 59 Temple Place - Suite 330," << endl;
  cout << "Boston, MA 02111-1307, USA." << endl;
  cout << endl << endl;
  cout << "Report bugs to doug@dougsparling.com." << endl;
}
